doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod

```diff
@@ -0,0 +1,149 @@
+
+In this mode the digest, key, salt and info values must be set before a key is
+derived or an error occurs.
+
+=item EVP_KDF_HKDEF_MODE_EXTRACT_ONLY
+
+In this mode calling L<EVP_KDF_CTX_derive(3)> will just perform the extract
+operation. The value returned will be the intermediate fixed-length pseudorandom
+key K.
+
+The digest, key and salt values must be set before a key is derived or an
+error occurs.
+
+=item EVP_KDF_HKDEF_MODE_EXPAND_ONLY
+
+In this mode calling L<EVP_KDF_CTX_derive(3)> will just perform the expand
+operation. The input key should be set to the intermediate fixed-length
+pseudorandom key K returned from a previous extract operation.
+
+The digest, key and info values must be set before a key is derived or an
+error occurs.
+
+=back
+
+=head1 STRING CTRLS
+
+HKDF also supports string based control operations via
+L<EVP_KDF_CTX_ctrl_str(3)>.
+The B<type> parameter "md" uses the supplied B<value> as the name of the digest
+algorithm to use.
+The B<type> parameter "mode" uses the values "EXTRACT_AND_EXPAND",
+"EXTRACT_ONLY" and "EXPAND_ONLY" to determine the mode to use.
+The B<type> parameters "salt", "key" and "info" use the supplied B<value>
+parameter as a B<seed>, B<key> or B<info> value.
+The names "hexsalt", "hexkey" and "hexinfo" are similar except they take a hex
+string which is converted to binary.
+
+=head1 NOTES
+
+All these functions are implemented as macros.
+
+A context for HKDF can be obtained by calling:
+
+ EVP_KDF_CTX *kctx = EVP_KDF_CTX_new_id(EVP_KDF_HKDF);
+
+The total length of the info buffer cannot exceed 1024 bytes in length: this
+should be more than enough for any normal use of HKDF.
+
+The output length of an HKDF expand operation is specified via the length
+parameter to the L<EVP_KDF_CTX_derive(3)> function.  Since the HKDF output
+length is variable, passing a B<NULL> buffer as a means to obtain the requisite
+length is not meaningful with HKDF in any mode that performs an expand
+operation.  Instead, the caller must allocate a buffer of the desired length,
+and pass that buffer to L<EVP_KDF_CTX_derive(3)> along with (a pointer
+initialized to) the desired length. Passing a B<NULL> buffer to obtain the
+length is allowed when using EVP_KDF_HKDEF_MODE_EXTRACT_ONLY.
+
+=head1 RETURN VALUES
+
+All these functions return 1 for success and 0 or a negative value for failure.
+In particular a return value of -2 indicates the operation is not supported by
+the KDF algorithm.
+
+=head1 EXAMPLE
+
+This example derives 10 bytes using SHA-256 with the secret key "secret",
+salt value "salt" and info value "label":
+
+ EVP_KDF_CTX *kctx;
+ unsigned char out[10];
+ size_t outlen = sizeof(out);
+ kctx = EVP_KDF_CTX_new_id(EVP_KDF_HKDF);
+
+ if (EVP_KDF_CTX_set_md(kctx, EVP_sha256()) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_set1_salt(kctx, "salt", 4) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_set1_key(kctx, "secret", 6) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_add1_hkdf_info(kctx, "label", 5) <= 0)
```

Comment:
Should `EVP_KDF_CTX_add1_hkdf_info` be renamed to something containing "label"? e.g. `EVP_KDF_CTX_add1_label`

Reply:
That would be better! "label" is used in nearby locations with a different meaning. e.g. https://github.com/openssl/openssl/pull/6674/files#diff-4e6407b0ec75b41bf1eb29a8ef274bdbR28
```
Ctrl.label = seed:key expansion
```

Reply:
Well, `info` here is the name used in `HKDF` (RFC 5869), while `"label"` is just five bytes of sample data that makes reference to how it is used in TLS 1.3 (RFC 8446), a well-known user of `HKDF`.
Earlier TLS versions use `TLS1-PRF` and supply the label as the salt parameter to that KDF.